driver-core/src/main/com/mongodb/connection/CommandMessage.java

```diff
@@ -17,36 +17,172 @@
 
 package com.mongodb.connection;
 
+import com.mongodb.MongoNamespace;
+import com.mongodb.ReadPreference;
+import com.mongodb.internal.validator.MappedFieldNameValidator;
+import org.bson.BsonBinaryWriter;
+import org.bson.BsonDocument;
+import org.bson.BsonElement;
+import org.bson.BsonString;
+import org.bson.BsonWriter;
+import org.bson.FieldNameValidator;
+import org.bson.codecs.EncoderContext;
 import org.bson.io.BsonOutput;
 
-abstract class CommandMessage extends RequestMessage {
-    CommandMessage(final String collectionName, final OpCode opCode, final MessageSettings settings) {
-        super(collectionName, opCode, settings);
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.mongodb.ReadPreference.primary;
+import static com.mongodb.connection.BsonWriterHelper.writePayload;
+
+/**
+ * A command message that uses OP_MSG or OP_QUERY to send the command.
+ */
+final class CommandMessage extends RequestMessage {
+    private final MongoNamespace namespace;
+    private final BsonDocument command;
+    private final FieldNameValidator commandFieldNameValidator;
+    private final ReadPreference readPreference;
+    private final SplittablePayload payload;
+    private final FieldNameValidator payloadFieldNameValidator;
+    private final boolean responseExpected;
+
+    CommandMessage(final MongoNamespace namespace, final BsonDocument command, final FieldNameValidator commandFieldNameValidator,
+                   final ReadPreference readPreference, final MessageSettings settings) {
+        this(namespace, command, commandFieldNameValidator, readPreference, settings, true, null, null);
     }
 
-    abstract boolean isResponseExpected();
+    CommandMessage(final MongoNamespace namespace, final BsonDocument command, final FieldNameValidator commandFieldNameValidator,
+                   final ReadPreference readPreference, final MessageSettings settings, final boolean responseExpected,
+                   final SplittablePayload payload, final FieldNameValidator payloadFieldNameValidator) {
+        super(namespace.getFullName(), getOpCode(settings), settings);
+        this.namespace = namespace;
+        this.command = command;
+        this.commandFieldNameValidator = commandFieldNameValidator;
+        this.responseExpected = responseExpected;
+        this.payload = payload;
+        this.payloadFieldNameValidator = payloadFieldNameValidator;
+        this.readPreference = readPreference;
+    }
+
+    boolean containsPayload() {
+        return payload != null;
+    }
 
-    abstract EncodingMetadata encodeMessageBodyWithMetadata(BsonOutput bsonOutput, SessionContext sessionContext);
+    boolean isResponseExpected() {
+        return !useOpMsg() || responseExpected;
```

Comment:
I'm not sure how to test this.

Reply:
I was thinking that this would have caused a weird test failure, where a response was on the wire but not read, and then the next read would pick it up instead of the one it was expecting.  The fact that that didn't happen suggests that there is no test of an unacknowledged write that had to be split.

Reply:
Done.